File: shared/types/workbench.ts
import type {
  ApiCalendarEventResponse,
  CalendarEvent,
} from './calendar'
import type {
  ApiGlobalIndexQuoteResponse,
  ApiIndexQuoteResponse,
  ApiMarketReviewResponse,
  ApiMarketStatsResponse,
  ApiWatchlistQuoteItem,
  GlobalIndexQuote,
  IndexQuote,
  MarketReview,
  MarketStats,
  WatchlistQuote,
} from './market'
import type { ApiTelegraphResponse, TelegraphItem } from './telegraph'

/** 自选股概览行的 AI 分析状态：分组未开启 off / 已开启未生成 pending / 已生成 ready。 */
export type WorkbenchAiStatus = 'off' | 'pending' | 'ready'

/** 后端工作台自选股行。 */
export interface ApiWorkbenchWatchlistStock extends ApiWatchlistQuoteItem {
  aiStatus: WorkbenchAiStatus
  aiSummary: string | null
}

/** 后端工作台自选股分组。 */
export interface ApiWorkbenchWatchlistGroup {
  id: number
  name: string
  isDefault: boolean
  aiReviewEnabled: boolean
  items: ApiWorkbenchWatchlistStock[]
}

/** 工作台自选股行：行情 + AI 分析状态。 */
export interface WorkbenchWatchlistStock extends WatchlistQuote {
  aiStatus: WorkbenchAiStatus
  aiSummary: string | null
}

/** 工作台自选股分组。 */
export interface WorkbenchWatchlistGroup {
  id: number
  name: string
  isDefault: boolean
  aiReviewEnabled: boolean
  items: WorkbenchWatchlistStock[]
}

/** 后端板块资金动向卡单行（金额单位亿元）。 */
export interface ApiWorkbenchSectorFlowItem {
  sectorName: string
  changePct: number | null
  mainNetInflow: number | null
  topStockName: string | null
}

/** 板块资金动向卡单行（金额单位亿元）。 */
export interface WorkbenchSectorFlowItem {
  sectorName: string
  changePct: number | null
  mainNetInflow: number | null
  topStockName: string | null
}

/** 单个交易日的复盘生成结果。 */
export type ReviewDayStatus = 'success' | 'failed' | 'pending'

/** 后端近段交易日复盘结果行。 */
export interface ApiReviewDayStatus {
  tradeDate: string
  status: ReviewDayStatus
}

/** 视图模型：近段交易日复盘结果行。 */
export interface ReviewDayStatusItem {
  tradeDate: string
  status: ReviewDayStatus
}

/** 复盘整体状态：done 已生成 / pending 待生成 / failed 生成失败。 */
export type ReviewStatusState = 'done' | 'pending' | 'failed'

/** 后端复盘状态卡数据。 */
export interface ApiReviewStatus {
  status: ReviewStatusState
  tradeDate: string
  generatedAt: string | null
  durationSeconds: number | null
  plannedTime: string | null
  nextRunAt: string | null
  streakDays: number
  monthSuccessRate: number | null
  recentDays: ApiReviewDayStatus[]
}

/** 复盘状态卡数据。 */
export interface ReviewStatus {
  status: ReviewStatusState
  tradeDate: string
  generatedAt: string | null
  durationSeconds: number | null
  plannedTime: string | null
  nextRunAt: string | null
  streakDays: number
  monthSuccessRate: number | null
  recentDays: ReviewDayStatusItem[]
}

/** 后端采集引擎运行记录。 */
export interface ApiCollectorRunItem {
  taskName: string
  taskLabel: string
  source: string | null
  status: string
  startedAt: string | null
  finishedAt: string | null
  durationSeconds: number | null
  recordsCount: number | null
}

/** 采集引擎运行记录。 */
export interface CollectorRunItem {
  taskName: string
  taskLabel: string
  source: string | null
  status: string
  startedAt: string | null
  finishedAt: string | null
  durationSeconds: number | null
  recordsCount: number | null
}

/** 后端采集引擎计划运行项。 */
export interface ApiCollectorUpcomingItem {
  runAt: string
  taskName: string
  taskLabel: string
  source: string | null
}

/** 采集引擎计划运行项。 */
export interface CollectorUpcomingItem {
  runAt: string
  taskName: string
  taskLabel: string
  source: string | null
}

/** 后端采集引擎状态。 */
export interface ApiCollectorEngineStatus {
  isRunning: boolean
  running: ApiCollectorRunItem | null
  recentRuns: ApiCollectorRunItem[]
  upcoming: ApiCollectorUpcomingItem[]
}

/** 采集引擎状态。 */
export interface CollectorEngineStatus {
  isRunning: boolean
  running: CollectorRunItem | null
  recentRuns: CollectorRunItem[]
  upcoming: CollectorUpcomingItem[]
}

/** GET /workbench 原始聚合响应（camelCase wire）。 */
export interface ApiWorkbenchResponse {
  calendar: ApiCalendarEventResponse[]
  review: ApiMarketReviewResponse | null
  reviewStatus: ApiReviewStatus | null
  telegraph: ApiTelegraphResponse[]
  watchlistGroups: ApiWorkbenchWatchlistGroup[]
  indices: ApiIndexQuoteResponse[]
  stats: ApiMarketStatsResponse | null
  globalIndices: ApiGlobalIndexQuoteResponse[]
  sectorFlow: ApiWorkbenchSectorFlowItem[]
  collectorStatus: ApiCollectorEngineStatus | null
}

/** 工作台聚合的客户端视图模型。 */
export interface WorkbenchOverview {
  calendar: CalendarEvent[]
  review: MarketReview | null
  reviewStatus: ReviewStatus | null
  telegraph: TelegraphItem[]
  watchlistGroups: WorkbenchWatchlistGroup[]
  indices: IndexQuote[]
  stats: MarketStats | null
  globalIndices: GlobalIndexQuote[]
  sectorFlow: WorkbenchSectorFlowItem[]
  collectorStatus: CollectorEngineStatus | null
}

File: backend/app/schemas/workbench.py
"""工作台聚合响应的 Pydantic schemas。"""

from datetime import date, datetime
from typing import Literal

from app.schemas.base import CamelModel
from app.schemas.calendar import NewsCalendarEventResponse
from app.schemas.market import (
    GlobalIndexQuoteResponse,
    IndexQuoteResponse,
    MarketReviewResponse,
    MarketStatsResponse,
    WatchlistQuoteItem,
)
from app.schemas.telegraph import TelegraphResponse


class WorkbenchWatchlistStock(WatchlistQuoteItem):
    """自选股概览行：行情快照 + 所属分组开启 AI 复盘时的分析状态与摘要。"""

    ai_status: Literal["off", "pending", "ready"] = "off"
    ai_summary: str | None = None


class WorkbenchWatchlistGroup(CamelModel):
    """工作台自选股概览的分组容器。"""

    id: int
    name: str
    is_default: bool = False
    ai_review_enabled: bool = False
    items: list[WorkbenchWatchlistStock] = []


class SectorFlowItem(CamelModel):
    """板块资金动向卡单行：最新交易日主力净流入排行（金额单位亿元）。"""

    sector_name: str
    change_pct: float | None = None
    main_net_inflow: float | None = None
    top_stock_name: str | None = None


class ReviewDayStatus(CamelModel):
    """近段交易日单日复盘生成结果。"""

    trade_date: date
    status: Literal["success", "failed", "pending"]


class ReviewStatusResponse(CamelModel):
    """复盘状态卡数据：做没做 / 何时做 / 做得怎样（正文不在工作台展示）。"""

    status: Literal["done", "pending", "failed"]
    trade_date: date
    generated_at: datetime | None = None
    duration_seconds: int | None = None
    planned_time: str | None = None
    next_run_at: datetime | None = None
    streak_days: int = 0
    month_success_rate: float | None = None
    recent_days: list[ReviewDayStatus] = []


class CollectorRunItem(CamelModel):
    """采集引擎最近一条运行记录。"""

    task_name: str
    task_label: str
    source: str | None = None
    status: str
    started_at: datetime | None = None
    finished_at: datetime | None = None
    duration_seconds: int | None = None
    records_count: int | None = None


class CollectorUpcomingItem(CamelModel):
    """采集引擎未来计划中的一次运行。"""

    run_at: datetime
    task_name: str
    task_label: str
    source: str | None = None


class CollectorStatusResponse(CamelModel):
    """采集引擎状态卡数据：是否在跑 / 接下来半天跑什么 / 最近跑得怎样。"""

    is_running: bool = False
    running: CollectorRunItem | None = None
    recent_runs: list[CollectorRunItem] = []
    upcoming: list[CollectorUpcomingItem] = []


class WorkbenchResponse(CamelModel):
    """工作台聚合数据；单模块降级时对应字段为空态而非整体报错。"""

    calendar: list[NewsCalendarEventResponse] = []
    review: MarketReviewResponse | None = None
    review_status: ReviewStatusResponse | None = None
    telegraph: list[TelegraphResponse] = []
    watchlist_groups: list[WorkbenchWatchlistGroup] = []
    indices: list[IndexQuoteResponse] = []
    stats: MarketStatsResponse | None = None
    global_indices: list[GlobalIndexQuoteResponse] = []
    sector_flow: list[SectorFlowItem] = []
    collector_status: CollectorStatusResponse | None = None
